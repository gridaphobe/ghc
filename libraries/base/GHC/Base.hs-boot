{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Base where

import GHC.Types

(++) :: [a] -> [a] -> [a]

type String = [Char]
