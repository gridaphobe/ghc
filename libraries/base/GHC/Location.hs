{-# LANGUAGE RecordWildCards #-}
module GHC.Location
  ( CallStack
  , getCallStack

  , SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol

  -- * Pretty printing
  , showCallStack
  , showSrcLoc
  ) where

--------------------------------------------------------------------------------
-- Source Locations
--------------------------------------------------------------------------------

-- | A CallStack is comprised of at least one 'SrcLoc'.
data CallStack = CallStack { getCallStack :: [SrcLoc] }
  deriving (Show, Eq)

showCallStack :: CallStack -> String
showCallStack (CallStack (root:rest))
  = unlines (showSrcLoc root : map (indent . showSrcLoc) rest)
  where
  indent l = "  " ++ l
showCallStack _ = error "CallStack cannot be empty!"

-- | A single location in the source code.
data SrcLoc = SrcLoc
  { srcLocPackage   :: String
  , srcLocModule    :: String
  , srcLocFile      :: String
  , srcLocStartLine :: Int
  , srcLocStartCol  :: Int
  , srcLocEndLine   :: Int
  , srcLocEndCol    :: Int
  } deriving (Show, Eq)

showSrcLoc :: SrcLoc -> String
showSrcLoc SrcLoc {..}
  = concat [ srcLocFile, ":"
           , show srcLocStartLine, ":"
           , show srcLocStartCol, " in "
           , srcLocPackage, ":", srcLocModule
           ]
