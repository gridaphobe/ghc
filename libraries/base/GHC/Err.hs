{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude, MagicHash #-}
{-# OPTIONS_HADDOCK hide #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Err
-- Copyright   :  (c) The University of Glasgow, 1994-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- The "GHC.Err" module defines the code for the wired-in error functions,
-- which have a special type in the compiler (with \"open tyvars\").
--
-- We cannot define these functions in a module where they might be used
-- (e.g., "GHC.Base"), because the magical wired-in type will get confused
-- with what the typechecker figures out.
--
-----------------------------------------------------------------------------

module GHC.Err( error ) where
import GHC.CString ()
import GHC.Types
import GHC.Prim
import GHC.Integer ()   -- Make sure Integer is compiled first
                        -- because GHC depends on it in a wired-in way
                        -- so the build system doesn't see the dependency
import {-# SOURCE #-} GHC.Exception( errorCallException )

-- | 'error' stops execution and displays an error message.
error :: [Char] -> CallStack -> a
error s stk = raise# (errorCallException s stk)
