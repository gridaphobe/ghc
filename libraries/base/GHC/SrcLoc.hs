{-# LANGUAGE NoImplicitPrelude, RecordWildCards #-}

-- | @since 4.8.2.0
module GHC.SrcLoc
  ( SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol

  -- * Pretty printing
  , showSrcLoc
  ) where

import GHC.Base
import GHC.Show

-- | Pretty print 'SrcLoc'
--
-- @since 4.8.2.0
showSrcLoc :: SrcLoc -> [Char]
showSrcLoc SrcLoc {..}
  = foldr (++) ""
      [ srcLocFile, ":"
      , show srcLocStartLine, ":"
      , show srcLocStartCol, " in "
      , srcLocPackage, ":", srcLocModule
      ]

