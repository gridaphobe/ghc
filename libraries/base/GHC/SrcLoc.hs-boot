{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.SrcLoc
  ( SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol
  ) where

import GHC.Types( Char, Int )

-- | A single location in the source code.
data SrcLoc = SrcLoc
  { srcLocPackage   :: [Char]
  , srcLocModule    :: [Char]
  , srcLocFile      :: [Char]
  , srcLocStartLine :: Int
  , srcLocStartCol  :: Int
  , srcLocEndLine   :: Int
  , srcLocEndCol    :: Int
  }
