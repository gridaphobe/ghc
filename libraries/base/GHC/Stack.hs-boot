-- This SOURCE-imported hs-boot module cuts a dependency loop:
--
--          GHC.Stack
-- imports  GHC.Base
-- imports  GHC.Err
-- imports  {-# SOURCE #-} GHC.Stack

{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Stack ( CallStack, showCallStack ) where

import {-# SOURCE #-} GHC.SrcLoc ( SrcLoc )
import GHC.Types ( Char )

data CallStack = CallStack { getCallStack :: [([Char], SrcLoc)] }

showCallStack :: CallStack -> [Char]
