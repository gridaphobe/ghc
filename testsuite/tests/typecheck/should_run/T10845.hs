{-# LANGUAGE ImplicitParams #-}
{-# OPTIONS_GHC -dcore-lint #-}

import GHC.Exception
import GHC.Stack

f1 :: (?loc :: CallStack) => CallStack
f1 = let y = (?loc :: CallStack)
     in y

f2 :: (?loc :: CallStack) => CallStack
f2 = let y x = ?loc
     in y 0

main :: IO ()
main = do putStrLn $ showCallStack f1
          putStrLn $ showCallStack f2
