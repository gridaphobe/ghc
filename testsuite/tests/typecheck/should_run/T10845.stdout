CallStack:
  ?loc, called at T10845.hs:8:15 in main:Main
  f1, called at T10845.hs:29:36 in main:Main
CallStack:
  ?loc, called at T10845.hs:12:17 in main:Main
  y, called at T10845.hs:13:9 in main:Main
  f2, called at T10845.hs:30:36 in main:Main
CallStack:
  y, called at T10845.hs:21:9 in main:Main
  f3, called at T10845.hs:31:36 in main:Main
CallStack:
  ?loc, called at T10845.hs:24:44 in main:Main
  y, called at T10845.hs:26:9 in main:Main
  f4, called at T10845.hs:32:11 in main:Main
CallStack:
  ?loc, called at T10845.hs:24:44 in main:Main
  y, called at T10845.hs:26:9 in main:Main
  f4, called at T10845.hs:32:11 in main:Main
